src/Tracer.php

```diff
@@ -0,0 +1,203 @@
+            $this->exerciseDir . '/',
+            '',
+            $phpUnitMessage
+        );
+        $phpUnitMessage = $testMethod->nameWithClass() . "\n" . $phpUnitMessage;
+
+        $this->result['tests'][] = new Result(
+            $testMethod->testDox()->prettifiedMethodName(),
+            'error',
+            $this->methodCode($testMethod),
+            $this->taskId($testMethod),
+            '',
+            $phpUnitMessage,
+        );
+    }
+
+    private function addBeforeFirstTestMethodErrored(BeforeFirstTestMethodErrored $event): void
+    {
+        $phpUnitMessage = \trim($event->throwable()->asString());
+        $phpUnitMessage = \str_replace(
+            $this->exerciseDir . '/',
+            '',
+            $phpUnitMessage
+        );
+
+        $this->result['status'] = 'error';
+        $this->result['message'] = $phpUnitMessage;
+    }
+
+    private function addTestOutput(PrintedUnexpectedOutput $event): void
+    {
+        // This must rely on the sequence of events!
+
+        /** @var Result */
+        $lastTest = $this->result['tests'][\array_key_last($this->result['tests'])];
+        $lastTest->setUserOutput($event->output());
+    }
+
+    private function saveResults(): void
+    {
+        foreach ($this->result['tests'] as $result) {
+            if ($result->isFailed() || $result->isErrored()) {
+                $this->result['status'] = 'fail';
+            }
+        }
+
+        \file_put_contents(
+            $this->outFileName,
+            \json_encode($this->result) . "\n",
+            // \json_encode($this->result, JSON_PRETTY_PRINT) . "\n",
+        );
+    }
+
+    private function methodCode(TestMethod $testMethod): string
+    {
+        $reflectionClass = new ReflectionClass($testMethod->className());
+        $reflectionMethod = $reflectionClass->getMethod($testMethod->methodName());
+
+        // Line numbers are 1-based, array index is 0-based.
+        // Reflections start line is the function declaration, end line is
+        // closing curly bracket.
+        // We use PSR-12, which makes line based code extraction problematic
+        // (function parameters may be on multiple lines). But we have 99% of
+        // code starting on second line after function declaration, and the
+        // closing bracket will be on the line after the last code line.
+        $start = $reflectionMethod->getStartLine() - 1 + 2;
+        $end = $reflectionMethod->getEndLine() - 1 - 1;
+
+        $codeLines = \array_filter(
+            \file($reflectionMethod->getFileName()),
+            static fn ($index) => $index >= $start && $index <= $end,
+            ARRAY_FILTER_USE_KEY,
+        );
+
+        // Unindent lines 2 levels of 4 spaces each (if possible)
+        $codeLines = \array_map(
+            fn ($line) => \str_starts_with($line, '        ')
+                ? \substr($line, 2 * 4)
+                : $line
+                ,
+            $codeLines,
+        );
```

Comment:
Would it make sense to get the indentation of the first line then unindent all of them by the same indentation?

Or alternatively do one pass to find the minimum indentation and unindent all the lines by this amount.

Reply:
I didn't want to spend a lot of time / effort on this. We use PSR-12, with tooling to ensure the indentation. What this does is prevent empty lines from being removed. That's the only reason I test for enough spaces to remove.